liquibase-standard/src/main/java/liquibase/snapshot/SnapshotGeneratorChain.java

```diff
@@ -40,34 +42,31 @@ public <T extends DatabaseObject> T snapshot(T example, DatabaseSnapshot snapsho
             return null;
         }
 
-        SnapshotGenerator next = getNextValidGenerator();
-
-        if (next == null) {
-            return null;
-        }
-
-        T obj = next.snapshot(example, snapshot, this);
-        if ((obj != null) && (obj.getSnapshotId() == null)) {
-            obj.setSnapshotId(snapshotIdService.generateId());
-        }
-        return obj;
-    }
-
-    public SnapshotGenerator getNextValidGenerator() {
         if (snapshotGenerators == null) {
             return null;
         }
 
-        if (!snapshotGenerators.hasNext()) {
-            return null;
-        }
-
-        SnapshotGenerator next = snapshotGenerators.next();
-        for (Class<? extends SnapshotGenerator> removedGenerator : replacedGenerators) {
-            if (removedGenerator.equals(next.getClass())) {
-                return getNextValidGenerator();
+        T result = null;
+        boolean resultInitialized = false;
+        while (snapshotGenerators.hasNext()) {
+            SnapshotGenerator generator = snapshotGenerators.next();
+            if (replacedGenerators.stream()
+                    .anyMatch(Predicate.isEqual(generator.getClass()))) {
+                continue;
+            }
+            T object = generator.snapshot(example, snapshot, this);
+            if ((object != null) && (object.getSnapshotId() == null)) {
+                object.setSnapshotId(snapshotIdService.generateId());
+            }
+            if (resultInitialized && result != object) {
+                Scope.getCurrentScope().getLog(getClass())
```

Comment:
I actually need to double-check this, not sure it makes sense

Reply:
I was also tempted to throw here
ideally subsequent snapshot generators should only modify instance of database objects (mostly by adding nested attributes), not return new ones

Reply:
ok, I tried the exception approach and it works out quite well in my case. If I forget to replace the built-in catalog snapshot generator, I get this error:

```java
liquibase.exception.DatabaseException: Snapshot generator liquibase.snapshot.jvm.CatalogSnapshotGenerator has returned a different reference from the previous generator liquibase.ext.neo4j.snapshot.TypeSnapshotGeneratorNeo4j.
	Snapshot object was: class liquibase.structure.core.Catalog@1202508091, it is now: class liquibase.structure.core.Catalog@1871617286.
	Consider adding one of the generator to the result of the other's implementation of liquibase.snapshot.SnapshotGenerator#replaces.
```

Reply:
added in https://github.com/liquibase/liquibase/pull/5619/commits/e4fcf69684b3738c02c608e525594ed799a3782e